feat(plan): add plan section support

flotter opened the pull request:
Add support for Pebble Plan extensions to allow managers to extend the existing Pebble Plan schema.

```
// LayerSectionExtension allows the plan layer schema to be extended without
// adding centralised schema knowledge to the plan library.
type LayerSectionExtension interface {
	// ParseSection returns a newly allocated concrete type containing the
	// unmarshalled section content.
	ParseSection(data yaml.Node) (LayerSection, error)

	// CombineSections returns a newly allocated concrete type containing the
	// result of combining the supplied sections in order.
	CombineSections(sections ...LayerSection) (LayerSection, error)

	// ValidatePlan takes the complete plan as input, and allows the
	// extension to validate the plan. This can be used for cross section
	// dependency validation.
	ValidatePlan(plan *Plan) error
}

type LayerSection interface {
	// Ask the section to validate itself.
	Validate() error

	// Returns true if the section is empty.
	IsZero() bool
}
```

flotter committed:
feat(plan): add plan section support

AnneCYH committed:
docs: how to manage service dependencies (#457)

IronCore864 committed:
feat(plan): use lanes to start services with dependencies (#437)

Put services that depend on each other (requires) in the same lane to start/stop in parallel.

flotter commented on internals/plan/suite_test.go:
Moved test suite setup code to ```package_test.go``` as is becoming the new norm in Pebble.

paul-rodriguez commented on internals/plan/plan.go:
```diff
@@ -32,6 +32,32 @@ import (
+type LayerSectionExtension interface {
+	// ParseSection creates a new layer section containing the unmarshalled
+	// yaml.Node, and any additional section specifics it wishes to apply
+	// to the backing type. A nil LayerSection returned ensures that this
+	// section will be omitted by the caller.
```
It's a bit strange to comment on the behavior of the caller in these functions.

I would expect the same idea to be formulated as "The return value is optional and a nil value signifies an absence of LayerSection".

flotter replied:
I've removed the caller wording. However, as the consumer of the interface, I do need to inform the implementer how to behave in order to achieve the expected results.

```https://pkg.go.dev/io#Reader```

For example, for ```io.Reader```:

```Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.```

paul-rodriguez commented on internals/plan/plan.go:
```diff
@@ -42,11 +68,33 @@ const (
 )
 
+// layerExtensions keeps a map of registered extensions.
+var layerExtensions = map[string]LayerSectionExtension{}
```
This feels bad. I think it's not that unrealistic to pass the extensions along the call chain from PlanManager to the Plan and into each Layer. Obviously this would be an invasive refactor.

flotter replied:
Yes, you are right this is a bit unusual.

I think the reason this was preferred was because the ```Plan``` library type is used as an immutable snapshot of the current plan, so its short lived. Throughout the lifetime of Pebble, new instances of ```Plan{}``` is created with the latest set of layers, and a combined plan. The extensions on the other hand stays the same throughout the lifetime of the Pebble process.

Instead of the Plan Manager receiving the extensions, and configuring the plan library once, as you say, the Plan Manager could ensure every ```Plan``` instance has a pointer to the extensions. I will keep this open and discuss the option with Gustavo / Ben.

paul-rodriguez replied:
> Instead of the Plan Manager receiving the extensions, and configuring the plan library once, as you say, the Plan Manager could ensure every Plan instance has a pointer to the extensions.

Yeah this is more or less what I was thinking

flotter replied:
Going to go ahead with this approach, following a quick revisit (chat with CTO). Its not perfect, but there is not an obvious solution that is only scoped within this package. The alternatives break the common understand of the lifetime of the plan instance. Resolving.

paul-rodriguez commented on internals/plan/plan.go:
```diff
@@ -643,6 +705,30 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
+				sections = append(sections, section)
+			}
+		}
+		// Deliberately do not expose the zero section condition to the extension.
+		// For now, the result of combining nothing must result in an omitted section.
+		if len(sections) > 0 {
```
Not sure why this is checked here rather than inside `extension.CombineSections`

flotter replied:
As the comment explains, I do not want to give the implementer the option to return something, even if no sections are provided as input. For now, I want to restrict this, and prevent section implementations (i.e. netstate) from making a call how to behave in this case.

thp-canonical committed:
feat(daemon): allow overriding layers directory in daemon.Options (#456)

This allows having the layers directory in a different directory than
the "layers" sub-directory inside the pebble directory.

The defaults are unchanged, but one can pass an optional non-empty
string in `daemon.Options{LayersDir: ...}` when calling `daemon.New()`
to use a custom location for the layers.

This also removes the need to pass `pebbleDir` into `plan` and
`planstate`, as it really only needs the `layersDir`.

flotter commented on internals/plan/plan.go:
```diff
@@ -42,11 +67,33 @@ const (
+// Section retrieves a section from the plan. Returns nil if
+// the field does not exist.
+func (p *Plan) Section(field string) LayerSection {
+	return p.Sections[field]
```
Should we support an error if the field was not registered in the first place ?

anpep replied:
That's probably the only error we could find so either we return `nil` (my personal preference) or we return `(section LayerSection, ok bool)`

flotter replied:
Revamped this. Resolving.

flotter commented on internals/plan/extensions_test.go:
```diff
@@ -0,0 +1,646 @@
+		} else {
+			if planTest.result != nil {
+				// Verify "x-field" data.
+				x := p.Section(xField).(*xSection)
```
If the section does not exist, the type assertion will fail. This means consumer code will first have to test for nil and only proceed with type assertion if it is not nil.

flotter replied:
Revamped this. Resolving.

anpep commented on internals/plan/plan.go:
```diff
@@ -57,6 +104,19 @@ type Layer struct {
+	layer.Sections[field] = section
+}
+
+// Section retrieves a layer section from a layer. Returns nil if
+// the field does not exist.
+func (layer *Layer) Section(field string) LayerSection {
+	return layer.Sections[field]
 }
```
`Layer.Sections` being unexported would justify having a `Section(string)` getter. However, this method is equivalent to just indexing `layer.Sections`. Also, `layer.Sections[field]` reads obvious and clear, why bother adding a getter?

flotter replied:
I've removed this, essentially inlining it in placed its used.

flotter replied:
Resolving.

anpep commented on internals/plan/plan.go:
```diff
@@ -1020,19 +1121,80 @@ func (p *Plan) checkCycles() error {
 	}
+
+	for field, section := range layerSections {
+		switch field {
+		case "summary", "description", "services", "checks", "log-targets":
```
nit (so we don't have to keep an up-to-date slice of fields twice):
```suggestion
		if _, builtin := builtinSections[field]; builtin {
				// ... case "summary", "description", "services", "checks", "log-targets":
		} else {
				// ... default:
		}
```

flotter replied:
Fixed this. Great proposal. Resolving.

anpep commented on internals/plan/plan.go:
nit: you updated the copyright notice in `plan_test.go` but not in `plan.go`?

flotter replied:
Fixed. Resolving.

paul-rodriguez commented on internals/plan/plan.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2021 Canonical Ltd
+// Copyright (c) 2024 Canonical Ltd
```
This should become a range right?

flotter replied:
I've seen people in snapd simply updating to latest, including in Pebble, so I think this is OK. Resolving.

acalinp committed:
feat(overlord): reversed order of manager stop (#447)

Currently managers are started and stopped in a FIFO (first in, first
out) order, but I would expect managers to behave more like a stack, in
that they are started and stopped in a LIFO (last in first out) order.
So the stopping of the managers should be done in reverse.

Co-authored-by: Ben Hoyt <ben.hoyt@canonical.com>

IronCore864 committed:
docs: improvements based on feedback from anne and daniele (#459)

Minor docs improvements.

IronCore864 committed:
docs: automate cmd reference docs generation (#458)

A script to help generate CLI CMD reference docs.

IronCore864 committed:
docs: restructure references (#461)

Structural changes to the reference docs.

IronCore864 committed:
docs: tweaks according to latest readme changes (#470)

[Recent changes made to the
README](https://github.com/canonical/pebble/commit/c330d5f0f37cd7c47b4ec3ce094b39e43fb3f6b4)
should be made to the doc site as well.

Closes #468.

dmitry-lyfar committed:
fix(overlord): send EOF to master pty in interactive mode (#466)

If the interactive mode is enforced in `pebble exec`, there are
scenarios like:

```go run ./cmd/pebble exec -i -- tee < ./HACKING.md```

that would prompt the server to shutdown the execution earlier than the whole output from 'tee' will be read and sent to the client. Hence, instead of closing the master pty descriptor when there is no more input from the client, send
Ctrl-D (EOF) to indicate to the process that there will be no more input.

Fixes #306.

flotter commented on internals/plan/extensions_test.go:
```diff
@@ -0,0 +1,709 @@
+			if planTest.result != nil {
+				// Verify "x-field" data.
+				var x *xSection
+				err = p.Section(xField, &x)
```
Having NOT an error should not result in an non-updated return type!!

flotter replied:
Fixed. Empty sections are now backed by an zero value type.

flotter commented on internals/plan/plan.go:
```diff
@@ -42,11 +67,57 @@ const (
+
+// ResetExtensions should only be used during testing to reset the registered
+// extensions in between different tests.
+func ResetExtensions() {
```
UnregisterExtension(field string, ext LayerSectionExtension)

flotter replied:
Applied.

IronCore864 committed:
docs: update tutorial according to review (#471)

Minor changes to the tutorial according to Anne's suggestion.

IronCore864 committed:
docs: add explanations for how to manage service dependencies (#440)

A how-to guide on managing service dependencies.

benhoyt committed:
fix(cmdstate): don't store execSetup and environment in state (#478)

Per #411, the environment (and other task data) can get quite large for
exec objects. Given that exec commands are one-shot and only relevant
for the current run of Pebble, there's no need to persist them.

Note that none of this data is in "api-data" so it's not accessible via
the API at all right now, and so this is a non-breaking change.

Later if we want to make exec tasks a bit more introspectable we can add
the command line or other fields to "api-data" (but we shouldn't add the
entire environment).

Fixes #411.

IronCore864 committed:
chore: minor improvement to docs generation (#477)

Minor fix to the automated doc generation.

IronCore864 committed:
fix(overlord): pebble hangs with a read-only file system when starting the daemon (#481)

Fix an issue where the Pebble daemon fails to start with a read-only file system.

IronCore864 committed:
docs: enrich the landing page with basic concepts (#472)

Improve the definition of Pebble and add fundamentals.
Co-authored-by: Ben Hoyt <benhoyt@gmail.com>

benhoyt committed:
docs: add Identities documentation (#479)

This adds docs for Identities:

* Access control information under "API and clients"
* A "how to manage identities" guide
* And an "Identities" reference page to define/describe the concept

benhoyt committed:
docs: tweak per Anne's feedback (#484)

Clarify what "this" refers to now that *this* is in a different section.

IronCore864 committed:
docs: reduce readme (#483)

Reducing the readme to a minimum to get ready for launching our doc site.
---------
Co-authored-by: AnneCYH <69107893+AnneCYH@users.noreply.github.com>

thp-canonical committed:
fix(logger): Make MockLogger threadsafe (#441)

When using a `logger.MockLogger()`, a background goroutine calling
`logger.Noticef()` might race the underlying `bytes.Buffer` for
read/write access. To avoid that, limit the returned buffer type for
`MockLogger()` to just `fmt.Stringer` (and mutex-protect its reading
from the buffer) as well as mutex-protect `Write()` calls to the
underlying buffer.

To reproduce the race condition, run `go test -race ./internals/logger`
with `TestMockLoggerReadWriteThreadsafe` added, but with the unmodified
`MockLogger` implementation.

If more functions from `bytes.Buffer` are needed from the `MockLogger`
return value, the interface can be expanded with additional pass-through
lock-protected calls, but in all of Pebble, it seems only the
`.String()` function is used, so only that is exposed for now.

While `logger.Noticef()` calls are serialized by `logger.loggerLock`,
the same cannot be said for parallel access to the underlying
`bytes.Buffer` that is returned from `MockLogger` when calling
`logger.Noticef()` in one goroutine, and accessing the `bytes.Buffer`
from another goroutine.

IronCore864 committed:
docs: update in this documentation section (#485)

Update the "In this documentation" section on the main page.

flotter committed:
Make plan.Section() simpler - no reflection

flotter committed:
fix(overlord): allow ensure when state is available (#482)

Managers can explicitly call ```state.EnsureBefore()``` if they wish to
apply changes sooner than the scheduled 5min ensure overlord timer.

Although managers get a state instance at creation time, during
```overlord.New()```, it is currently forbidden to call ensure early on
before the ```Daemon``` started the ```overlord.Loop()```. The current
restriction stems from the fact that ```state.EnsureBefore()``` adjusts
the ensure timer expiry time, and the timer is only created when
```overlord.Loop()``` is called.

**The problem:**

Since the overlord ensure loop is only started later on during
```daemon.Start()```, a manager has no direct way to know when it's safe
to call Ensure. Managers currently have to use the first call to their
```Ensure()``` method as an indication the overlord loop was entered,
which is unnecessary manager boilerplate code. See
```overlord/checkstate/manager.go```.

Managers (and managers with dependencies on other managers), may want to
implement state changes during callbacks. These events can come from
outside the overlord framework (e.g. from the Linux kernel or early on
during manager ```StartUp```), meaning that handlers using state may be
triggered asynchronously from the Daemon startup sequence, and arrive
before the overlord Loop is started.

**Solution:**

Allow ```state.EnsureBefore()``` to be called before the timer exists.
Since an implicit ensure is performed on overlord loop entry, simply
returning while the timer does not exist is the same as saying an ensure
is already scheduled.

benhoyt committed:
refactor: implement warnings as notices (#486)

This deletes the legacy warnings code and re-implements `pebble
warnings` and `pebble okay` using notices as warnings (as we originally
intended). The `/v1/warnings` endpoints have been removed, and `pebble
warnings` now hits `/v1/notices?types=warning`.

On every request, instead of returning the count of warnings and the
latest timestamp, we only return the timestamp of the latest (most
recent last-repeated time). This way the server doesn't have to track
which notices have been okayed -- the client can compare against its
stored timestamp and see if the server-reported timestamp is later.

So there are a few breaking changes in this PR:

* `Client.WarningsSummary() (count int, timestamp time.Time)` is now
`Client.LatestWarningTime() time.Time`, as there's no count anymore
* `GET /v1/warnings` (list warnings) is gone: `pebble warnings` now uses
`GET /v1/notices?types=warning`
* `POST /v1/warnings` (okay warnings) are gone: `pebble okay` now just
updates the local state file, and doesn't use the API/client at all
* Similarly, `Client.Warnings()` and `Client.Okay()` are gone; use
`Client.Notices()` and maintain local client-side state, respectively
* `State.Warnf` is still around, but it's now just a helper that calls
`State.AddNotice` with the right type (warning) and key (message)

Fixes #475.

holmanb committed:
fix: initialize socketPath to prevent stale cache (#491)

`daemonSuite.socketPath` appears to be assigned outside of the test, but
only when it has a default value, which results in a stale `socketPath`
value in this test.

Running `go test -count=2 -v ./internals/daemon -check.v` with this fix
now passes.

Fixes #282.

flotter committed:
Review feedback group 1

flotter committed:
Review feedback group 2

flotter committed:
add test for layers content

IronCore864 committed:
test: improve osutil.mkdir test coverage (#492)

Added two more test cases to improve test coverage.

If the path already exists but it's not a directory, a `syscall.ENOTDIR`
error is expected.

tonyandrewmeyer committed:
ci: add a monthly (or on demand) TIOBE reporting workflow (#494)

Once a month, send the latest code quality (including coverage) data to
TIOBE for central storage.

---------

Co-authored-by: Ben Hoyt <benhoyt@gmail.com>

flotter committed:
enforce section marshal ordering

flotter committed:
Review feedback group 3

flotter committed:
Review feedback group 4

flotter committed:
Review feedback group 5

flotter committed:
Review feedback group 6

flotter committed:
Review feedback group 7

flotter committed:
tweak

flotter committed:
extension errors must provide full context

flotter committed:
remove redundant query

flotter committed:
tweak

flotter committed:
simplify YAML field changes detection

flotter committed:
naming tweak
